File: decker/tts.py
import json


def card_to_objectstate(index, edition, collector_number):
    """
    returns an object state dict for a TTS deck
    """
    card = index[edition][collector_number]
    face_url = card["image_uris"]["png"]

    # taken from https://github.com/jeandeaual/tts-deckconverter
    if card["layout"] in ["normal", "token", "split", "flip", "leveler", "emblem"]:
        back_url = "http://cloud-3.steamusercontent.com/ugc/998016607072060763/7AFEF2CE9E7A7DB735C93CF33CC4C378CBF4B20D/"
    elif card["layout"] == "planar":
        back_url = "http://cloud-3.steamusercontent.com/ugc/998016607072060000/1713AE8643632456D06F1BBA962C5514DD8CCC76/"
    elif card["layout"] == "scheme":
        back_url = "http://cloud-3.steamusercontent.com/ugc/998016607072055936/0598975AB8EC26E8956D84F9EC73BBE5754E6C80/"

    object_state = {
        "FaceURL": face_url,
        "BackURL": back_url,
        "NumWidth": 1,
        "NumHeight": 1,
        "BackIsHidden": True,
        "UniqueBack": False,
        "Type": 0
    }
    return object_state


def card_to_description(card):
    """
    returns a description given a card layout
    """
    if card["layout"] in ["normal", "token", "planar", "leveler", "scheme", "emblem"]:
        description = "[b]{}[/b]\n\n{}".format(card["type_line"], card["oracle_text"])
    elif card["layout"] in ["split", "flip"]:
        acc = []
        for face in card["card_faces"]:
            acc.append("[b]{}[/b]\n\n{}".format(face["type_line"], face["oracle_text"]))
        description = "\n\n--------------------------\n\n".join(acc)

    return description


def card_to_containedobject(index, edition, collector_number, card_id, object_state):
    """
    returns a contained object dict for a TTS deck
    """
    card = index[edition][collector_number]
    nickname = card["name"]
    description = card_to_description(card)
    idx = str(card_id//100)
    custom_deck = {idx: object_state}

    contained_object = {"Name": "Card",
                        "Transform": {"posX": 0.0, "posY": 0.0, "posZ": 0.0,
                                      "rotX": 0.0, "rotY": 180.0, "rotZ": 0.0,
                                      "scaleX": 1.1339285, "scaleY": 1.0, "scaleZ": 1.11125},
                        "Nickname": nickname,
                        "Description": description,
                        "GMNotes": "",
                        "ColorDiffuse": {"r": 0.71323913, "g": 0.71323913, "b": 0.71323913},
                        "Locked": False,
                        "Grid": True,
                        "Snap": True,
                        "IgnoreFoW": False,
                        "MeasureMovement": False,
                        "DragSelectable": True,
                        "Autoraise": True,
                        "Sticky": True,
                        "Tooltip": True,
                        "GridProjection": False,
                        "HideWhenFaceDown": True,
                        "Hands": True,
                        "CardID": card_id,
                        "SidewaysCard": False,
                        "CustomDeck": custom_deck,
                        "XmlUI": "",
                        "LuaScript": "",
                        "LuaScriptState": "",
                        "GUID": ""}

    return contained_object


def deck_to_ttscard(index, edition, collector_number):
    """
    returns a dict representing a tts card
    """
    object_state = card_to_objectstate(index, edition, collector_number)
    contained_object = card_to_containedobject(index, edition, collector_number, 100, object_state)
    ttscard = {"SaveName": "",
               "GameMode": "",
               "Gravity": 0.5,
               "PlayArea": 0.5,
               "Date": "",
               "Table": "",
               "Sky": "",
               "Note": "",
               "Rules": "",
               "LuaScript": "",
               "LuaScriptState": "",
               "XmlUI": "",
               "ObjectStates": [contained_object],
               "TabStates": {},
               "VersionNumber": ""}
    return ttscard


def deck_to_ttsdeck(index, deck):
    """
    returns a dict representing a tts deck
    """
    exploded_deck = []
    for deckline in deck:
        exploded_card = [(deckline["edition"], deckline["collector_number"])]*deckline["count"]
        exploded_deck.extend(exploded_card)

    if len(exploded_deck) == 1:
        (edition, collector_number) = exploded_deck[0]
        return deck_to_ttscard(index, edition, collector_number)

    custom_deck = {}
    contained_objects = []
    deck_ids = list(range(100, (len(exploded_deck)+1)*100, 100))
    for (card_id, (edition, collector_number)) in zip(deck_ids, exploded_deck):
        idx = str(card_id//100)
        object_state = card_to_objectstate(index, edition, collector_number)
        custom_deck[idx] = object_state
        contained_object = card_to_containedobject(index, edition, collector_number, card_id, object_state)
        contained_objects.append(contained_object)

    ttsdeck = {"SaveName": "",
               "GameMode": "",
               "Gravity": 0.5,
               "PlayArea": 0.5,
               "Date": "",
               "Table": "",
               "Sky": "",
               "Note": "",
               "Rules": "",
               "LuaScript": "",
               "LuaScriptState": "",
               "XmlUI": "",
               "ObjectStates": [{"Name": "Deck",
                                 "Transform": {"posX": 0.0, "posY": 0.0, "posZ": 0.0,
                                               "rotX": 0.0, "rotY": 180.0, "rotZ": 180.0,
                                               "scaleX": 1.1339285, "scaleY": 1.0, "scaleZ": 1.11125},
                                 "Nickname": "",
                                 "Description": "",
                                 "GMNotes": "",
                                 "ColorDiffuse": {"r": 0.71323913, "g": 0.71323913, "b": 0.71323913},
                                 "Locked": False,
                                 "Grid": True,
                                 "Snap": True,
                                 "IgnoreFoW": False,
                                 "MeasureMovement": False,
                                 "DragSelectable": True,
                                 "Autoraise": True,
                                 "Sticky": True,
                                 "Tooltip": True,
                                 "GridProjection": False,
                                 "HideWhenFaceDown": True,
                                 "Hands": False,
                                 "SidewaysCard": False,
                                 "DeckIDs": deck_ids,
                                 "CustomDeck": custom_deck,
                                 "XmlUI": "",
                                 "LuaScript": "",
                                 "LuaScriptState": "",
                                 "ContainedObjects": contained_objects,
                                 "GUID": ""}],
               "TabStates": {},
               "VersionNumber": ""}

    return ttsdeck

File: src/decker/tts.clj
(ns decker.tts
  (:require [clojure.java.io :as io]
            [cheshire.core :as json]
            [sanitize-filename.core :as sf]))


(def BACK-URL
  {"scheme" "http://cloud-3.steamusercontent.com/ugc/998016607072055936/0598975AB8EC26E8956D84F9EC73BBE5754E6C80/"
   "planar" "http://cloud-3.steamusercontent.com/ugc/998016607072060000/1713AE8643632456D06F1BBA962C5514DD8CCC76/"
   "normal" "http://cloud-3.steamusercontent.com/ugc/998016607072060763/7AFEF2CE9E7A7DB735C93CF33CC4C378CBF4B20D/"})


(defn card->ObjectState
  "returns an object state dict for a TTS deck,
  transformed? is used for double sided cards and returns an object for the reverse"
  ([card]
   (card->ObjectState card false))
  ([card transformed?]
   (case (:card/layout-category card)
     (:normal :split)
     {"FaceURL" (:card/png card)
      "BackURL" (get BACK-URL (:card/layout card) (BACK-URL "normal"))
      "NumHeight" 1
      "NumWidth" 1
      "BackIsHidden" true}

     :transform
     {"FaceURL" (get-in card [:card/faces (if transformed? 1 0) :png])
      "BackURL" (get BACK-URL (:card/layout card) (BACK-URL "normal"))
      "NumHeight" 1
      "NumWidth" 1
      "BackIsHidden" true})))


(defn card->Description
  "returns a description given a card layout
   transformed? is used for double sided cards and returns an object for the reverse"
  ([card]
   (card->Description card false))
  ([card transformed?]
   (case (:card/layout-category card)
     :normal
     (format "[b]%s[/b]\n\n%s" (:card/type-line card) (:card/oracle-text card))

     :split
     (format "[b]%s[/b]\n\n%s\n\n--------------------------\n\n[b]%s[/b]\n\n%s"
             (get-in card [:card/faces 0 :type-line])
             (get-in card [:card/faces 0 :oracle-text])
             (get-in card [:card/faces 1 :type-line])
             (get-in card [:card/faces 1 :oracle-text]))

     :double
     (let [face (if transformed? 1 0)]
       (format  "[b]%s[/b]\n\n%s"
                (get-in card [:card/faces face :type-line])
                (get-in card [:card/faces face :oracle-text]))))))


(defn tts-card->ContainedObject
  "returns a contained object dict for a TTS deck"
  ([tts-card]
   (tts-card->ContainedObject tts-card false))
  ([tts-card transformed?]
   (let [base-ContainedObject {"Name" "Card"
                               "Transform" {"posX" 0.0 "posY" 0.0 "posZ" 0.0
                                            "rotX" 0.0 "rotY" 180.0 "rotZ" 180.0
                                            "scaleX" 1.0 "scaleY" 1.0 "scaleZ" 1.0}
                               "Nickname" (get-in tts-card [:tts-card/card :card/name])
                               "Description" (card->Description (:tts-card/card tts-card) transformed?)}]
     (case (get-in tts-card [:tts-card/card :card/layout-category])
       (:normal :split)
       (assoc base-ContainedObject "CardID" (:tts-card/tts-id tts-card))

       :transform
       (if transformed?
         (assoc base-ContainedObject
                "CardID"
                (get-in tts-card [:tts-card/faces 1 :tts-id]))
         (assoc base-ContainedObject
                "CardID"
                (get-in tts-card [:tts-card/faces 0 :tts-id])
                "States"
                {"2" (tts-card->ContainedObject tts-card true)}))))))


(defn deck->tts-cards
  "Returns a list of tts-card in the deck, repeated for count
  the same card with multiple copies has the same id

  two faced cards come with 2 ids"
  [deck]
  (loop [tts-id 100 decklines (:deck/decklines deck) tts-cards []]
    (if (empty? decklines)
      tts-cards
      (let [deckline (first decklines)]
        (case (get-in deckline [:deckline/card :card/layout-category])
          (:normal :split)
          (recur (+ tts-id 100)
                 (rest decklines)
                 (->> #:tts-card{:tts-id tts-id
                                 :card (:deckline/card deckline)}
                      (repeat (:deckline/count deckline))
                      (concat tts-cards)))

          :transform
          (recur (+ tts-id 200)
                 (rest decklines)
                 (->> #:tts-card{:faces [{:tts-id tts-id}
                                         {:tts-id (+ tts-id 100)}]
                                 :card (:deckline/card deckline)}
                      (repeat (:deckline/count deckline))
                      (concat tts-cards))))))))


(defn tts-cards->CustomDeck
  "returns a map of card_id -> object_state"
  [tts-cards]
  (letfn [(reducer [acc tts-card]
            (case (get-in tts-card [:tts-card/card :card/layout-category])
              (:normal :split)
              (assoc acc
                     (str (/ (:tts-card/tts-id tts-card) 100))
                     (card->ObjectState (:tts-card/card tts-card)))

              :transform
              (assoc acc
                     (str (/ (get-in tts-card [:tts-card/faces 0 :tts-id]) 100))
                     (card->ObjectState (:tts-card/card tts-card))
                     (str (/ (get-in tts-card [:tts-card/faces 1 :tts-id]) 100))
                     (card->ObjectState (:tts-card/card tts-card) true))))]

    (reduce reducer {} tts-cards)))


(defn deck->tts-deck
  "returns a dict representing a tts deck"
  [deck]
  (let [tts-cards (deck->tts-cards deck)
        ContainedObjects (map tts-card->ContainedObject tts-cards)]
    {"ObjectStates" [{"Name" "DeckCustom"
                      "Nickname" (:deck/name deck)
                      "Description" (:deck/description deck)
                      "Transform" {"posX" 0.0 "posY" 1.0 "posZ" 0.0
                                   "rotX" 0.0 "rotY" 180.0 "rotZ" 180.0
                                   "scaleX" 1.0 "scaleY" 1.0 "scaleZ" 1.0}
                      "DeckIDs" (map #(get % "CardID") ContainedObjects)
                      "CustomDeck" (tts-cards->CustomDeck tts-cards)
                      "ContainedObjects" ContainedObjects}]}))


(defn write-tts-deck!
  [ttsdeck]
  (let [filename (format "%s.json" (sf/sanitize (get-in ttsdeck ["ObjectStates" 0 "Nickname"])))]
    (with-open [f (io/writer filename)]
      (spit filename (json/generate-string ttsdeck)))))
